snapshotter/app/service.go

```diff
@@ -129,7 +130,7 @@ func initSnapshotter(ctx context.Context, config config.Config) (snapshots.Snaps
 		return nil, err
 	}
 
-	newProxySnapshotterFunc := func(ctx context.Context, namespace string) (snapshots.Snapshotter, error) {
+	newProxySnapshotterFunc := func(ctx context.Context, namespace string) (*proxy.RemoteSnapshotter, error) {
```

Comment:
Can you explain your reasoning for this? It *feels* more correct to return the interface to me, but I don't have a strong reason for it.

Reply:
Sure. My reasoning for this is to couple the proxy snapshotter with its metrics proxy so that their lifetimes can be tracked together, i.e. when a the demux snapshotter invokes a cache miss and needs to use this `newProxySnapshotterFunc`, fetching that new proxy snapshotter will both fetch the snapshotter itself as well as initialize its metrics proxy server. When the demux snapshotter goes to evict a proxy snapshotter, this coupling allows for shutting down the metrisc proxy server as well.

Reply:
Maybe I missed it but I didn't see a place outside of the snapshotter where we need to expose the metrics proxy (e.g. shutting down the metrics proxy could be done inside the `RemoteSnapshotter`s close rather than requiring the cache to know about it).

However, on deeper reflection, I think the place we will need it is the eventual http discovery API so I think this is probably a change we need.

Reply:
Updated to implement the `Close` method of the `snapshots.Snapshotter` interface for `RemoteSnapshotter`. I'll be updating this PR with service discovery implementation so that we can consider those changes along with whether or not the cache should be caching the `RemoteSnapshotter` type or a `snapshots.Snapshotter` type - the latter of which might require some type assertion down the line.